### DISCUSSION TITLE
CropForeground not working

### DISCUSSION THREAD

**Discussion by oscarramirezs** [Project-MONAI/MONAI#6898] [Q&A]
Hello, 

I'm working on a project segmenting two bone from CT the images are  512x512. (The number of images can vary). 
My issue is that I need to create a nifti output for each bone without any background.
 However, when I do this, the coordinates of each bone are shifted compared to the original.

this is the output of my model 

                val_labels = [post_label(i) for i in decollate_batch(val_labels)]
                                          
                for i, c in enumerate(val_outputs):
                    label = ["bone1", "bone2"]
                       
                    bone1 = []
                    bone2 = []                                  
                  
                    bone1.append(c[1])
                    bone2.append(c[2])

                    bone1 = [np.stack(scapula, axis=0).astype(np.float16)]
                    result= np.array(bone1)

                    cropper = CropForeground(select_fn=lambda x: x > 1, channel_indices=None, margin=0, return_coords=True)
                    result_=cropper(result)

                    metadata=val_data["image_meta_dict"]  
                    
                    saver_h_2 = NiftiSaver(output_dir= "val_cut_test/", output_postfix='cut', output_ext=".nii")
                    saver_h_2.save_batch(result_[0], val_data["image_meta_dict"])       

can someone help me?

thanks 

Oscar

**Comment by KumoLiu** [comment]
Hi @oscarramirezs, thanks for your interest here.

If you want to crop both image and label, you could use `CropForegroundd` and set `keys=["image", "bone1"]`, `source_key="bone1"`.

https://github.com/Project-MONAI/MONAI/blob/2daabf917a1e6afbc738d22db602f4859672a90d/monai/transforms/croppad/dictionary.py#L709-L719

If you are still having issues, could you please provide a small piece of code and data so that I can reproduce your issue?
Thanks!

**Reply by oscarramirezs** [reply]
hi @KumoLiu 
Thank you for replaying,  but my issue is as follows:
 I have already trained the model that takes a ct 512x512 * (number of slices) and returns a mask of the same size with 0 for background and 1 for bone1. What I need is, after the inference, to receive a cropped mask that only contains the size of bone 1. 
My previous code does this well, but when I open the original CT and my cropped mask in 3DSlicer, the mask is not in the correct position relative to the original scanner.

**Reply by KumoLiu** [reply]
Hi @oscarramirezs, I think you need also to crop the original CT otherwise how can the cropped one have the same shape as the non-cropped one? Correct me if I misunderstood, thanks!

**Reply by oscarramirezs** [reply]
Actually, I don't need both on the same shape ,only the same position, and I need to cut only the output of the segmentation model. Here's an image below: I have the original CT and the original mask from the segmentation model. When I put them together in 3D Slicer, the mask is correctly aligned. But when I cut the mask and place it on 3D slice, the mask is shifted in relation to the original CT scan.

![image](https://github.com/Project-MONAI/MONAI/assets/52659349/f68a5ed7-d783-4d1e-8cd7-46bd1f5f131e)

Thank you for your help 

Oscar

**Reply by KumoLiu** [reply]
Hi @oscarramirezs, I didn't get the points about why you need cropping, I think you could directly filter the mask without cropping.
```
import numpy as np

a = np.array([
    [1, 2, 0],
    [0, 1, 2],
    [0, 1, 0],
])
a[a==1] = 0
```

**Reply by oscarramirezs** [reply]
Thank you. My goal is that the output mask goes to another AI model (not in MONAI). So, I need to reduce the size to decrease the weight  in megabytes for each patient. For example, an original mask can be 512x512x512, but the actual mask is only 45x45x45, with the rest of the matrix filled with zeros. Therefore, what I want is to have a much smaller matrix that retains all the mask information.

**Reply by KumoLiu** [reply]
Hi @oscarramirezs, the translation information is kept in the affine after `CropForeground`, so it can match within the ITK, here is the screenshot of the itk. Sorry I'm not familiar with the 3dslicer, but I think it could do the same thing.
<img width="1038" alt="Screen Shot 2023-08-24 at 18 17 27" src="https://github.com/Project-MONAI/MONAI/assets/55491388/4fc9a3ad-3d88-4092-b2d0-4d072598897f">

Thanks!

**Reply by oscarramirezs** [reply]
I dont knwo why in my case it doesn't work.

I have the val_output, how i should applie the crop? 
val_outputs = [post_pred(i) for i in decollate_batch(val_outputs)]


            val_labels = [post_label(i) for i in decollate_batch(val_labels)]
                                      
            for i, c in enumerate(val_outputs):
                label = ["bone1", "bone2"]
                   
                bone1 = []
                bone2 = []                                  
              
                bone1.append(c[1])
                bone2.append(c[2])

                bone1 = [np.stack(scapula, axis=0).astype(np.float16)]
                result= np.array(bone1)

                cropper = CropForeground(select_fn=lambda x: x > 1, channel_indices=None, margin=0, return_coords=True)
                result_=cropper(result)


                metadata=val_data["image_meta_dict"]  
                
                saver_h_2 = NiftiSaver(output_dir= "val_cut_test/", output_postfix='cut', output_ext=".nii")
                saver_h_2.save_batch(result_[0], val_data["image_meta_dict"])

**Reply by KumoLiu** [reply]
Hi @oscarramirezs, may I know your MONAI version? Could you please try with MONAI v1.2? You can check the updated affine by using `result_.affine`. We have updated to MetaTensor after v0.9.
You can find more info here. https://github.com/Project-MONAI/MONAI/wiki/MetaTensor-guide

Hope it helps, thanks!

**Reply by oscarramirezs** [reply]
Version de MONAI: 1.0.0. 
my script look fine?

**Reply by KumoLiu** [reply]
Since you didn't crop the image,  affine in `val_data["image_meta_dict"]` will not update, you need to use the updated meta information after `CropForeground`.

**Reply by oscarramirezs** [reply]
"I'm sorry, it's not working. I'm a bit new to this. Can I ask you the las question?"


Do you have a simple example code that reads a NIfTI, applies the crop function, and saves the new NIfTI, updating the affine matrix? Here's my code, but it's not working becouse, The CropForegroundd transformation does not return metadata in the form of a dictionary with "image_meta_dict" key. Instead, it returns a tuple containing the cropped image and the bounding box coordinates

import os
import glob
import numpy as np
from monai.transforms import Compose, CropForegroundd
from monai.data import NiftiSaver
import nibabel as nib

# Define the image directory
image_dir = r'C:\Users\Nifti1\val_cut_test\2EA840D22B5D_mimics_'
# Search for NIfTI image files
cut_images = sorted(glob.glob(os.path.join(image_dir, "*.nii")))

# Define the transformation
post_test = Compose([CropForegroundd(keys=["image"], source_key="image")])

# Initialize the NIfTI saver
saver = NiftiSaver(output_dir=image_dir, output_postfix='_cropped', output_ext=".nii")


nifti_img = nib.load(image_path)
img_data = nifti_img.get_fdata()

 
transformed_data = post_test({"image": img_data})["image"]

    
# Convert to a NumPy array of appropriate data type
transformed_data_np = transformed_data.detach().cpu().numpy().astype(np.float32)

# Save the transformed image
output_image_path = os.path.join(image_dir, os.path.basename(image_path).replace(".nii", "_cropped.nii"))
nib.save(nib.Nifti1Image(transformed_data_np, transformed_data.affine ), output_image_path)

print("Transformation and saving completed.")

**Reply by KumoLiu** [reply]
Hi @oscarramirezs, if you are using MONAI v1.0, I think you can directly use `Metatensor`.
Here is a simple example.
```
import monai.transforms as mt
data = {"img": img_path, "lbl": lbl_path}

trans = mt.Compose([
    mt.LoadImaged(keys=['img', 'lbl'], ensure_channel_first=True),
    mt.CropForegroundd(keys="lbl", source_key="lbl"),
])

out = trans(data)
print(out['lbl'].affine) # check whether the affine has been updated

mt.SaveImaged(keys='lbl', resample=False, output_dir='/workspace/')(out)
```

For more end-to-end examples, you can refer to the [tutorial repo](https://github.com/Project-MONAI/tutorials).
Hope it helps, thanks!
